crates/sui-indexer-alt-e2e-tests/tests/write_tests.rs

```diff
@@ -0,0 +1,374 @@
+                };
+                4
+            ])
+            .build()
+            .await;
+
+        // Unwrap since we know the URL should be valid.
+        let fullnode_rpc_url = Url::parse(onchain_cluster.rpc_url())?;
+
+        let cancel = CancellationToken::new();
+
+        let (rpc_handle, rpc_url) = set_up_rpc_server(fullnode_rpc_url, cancel.clone()).await;
+
+        Ok(Self {
+            onchain_cluster,
+            rpc_url,
+            rpc_handle,
+            client: Client::new(),
+            cancel,
+        })
+    }
+
+    /// Builds a simple transaction and returns the digest, tx bytes, and sigs to be used for testing.
+    async fn transfer_transaction(&self) -> anyhow::Result<(String, String, Vec<String>)> {
+        let addresses = self.onchain_cluster.wallet.get_addresses();
+
+        let recipient = addresses[1];
+        let tx = self
+            .onchain_cluster
+            .test_transaction_builder()
+            .await
+            .transfer_sui(Some(1_000), recipient)
+            .build();
+        let tx_digest = tx.digest().to_string();
+        let signed_tx = self.onchain_cluster.wallet.sign_transaction(&tx);
+        let (tx_bytes, sigs) = signed_tx.to_tx_bytes_and_signatures();
+        let tx_bytes = tx_bytes.encoded();
+        let sigs = sigs.iter().map(|sig| sig.encoded()).collect::<Vec<_>>();
+
+        Ok((tx_digest, tx_bytes, sigs))
+    }
+
+    /// Builds a transaction that would abort if called by a normal user.
+    async fn privileged_transaction(&self) -> anyhow::Result<(String, String, Vec<String>)> {
+        let tx: sui_types::transaction::TransactionData = self
+            .onchain_cluster
+            .test_transaction_builder()
+            .await
+            .call_request_remove_validator()
+            .build();
+        let tx_digest = tx.digest().to_string();
+        let signed_tx = self.onchain_cluster.wallet.sign_transaction(&tx);
+        let (tx_bytes, sigs) = signed_tx.to_tx_bytes_and_signatures();
+        let tx_bytes = tx_bytes.encoded();
+        let sigs = sigs.iter().map(|sig| sig.encoded()).collect::<Vec<_>>();
+
+        Ok((tx_digest, tx_bytes, sigs))
+    }
+
+    async fn execute_jsonrpc(&self, method: String, params: Value) -> anyhow::Result<Value> {
+        let query = json!({
+            "jsonrpc": "2.0",
+            "id": 1,
+            "method": method,
+            "params": params,
+        });
+
+        let response = self
+            .client
+            .post(self.rpc_url.clone())
+            .json(&query)
+            .send()
+            .await
+            .context("Request to JSON-RPC server failed")?;
+
+        let body: Value = response
+            .json()
+            .await
+            .context("Failed to parse JSON-RPC response")?;
+
+        Ok(body)
+    }
+
+    async fn stopped(self) {
+        self.cancel.cancel();
+        let _ = self.rpc_handle.await;
```

Comment:
I've noticed that this can hang for a long time (two minutes or more) due to `system_package_task` not shutting down, according to my printlns. I have not looked into it but do you have a good idea why this might happen? @amnn

Reply:
Interesting, is that happening in tests, or during normal running? Looking at the implementation of the task, the only hypothesis I have is that if the database is closed down before the system package task winds down, it could get stuck trying to get a connection to the DB which might have a long timeout.

I have not seen this happen in practice though so logs (from the system package task) would help make sense of what was going on.

Reply:
I don't know how test cluster cleans up or if it does cleaning at all. I don't see any code of this flavor in `test-cluster` crate.